activation/atxlistener.go

```diff
@@ -0,0 +1,71 @@
+package activation
+
+import (
+	"context"
+	"sync"
+
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"github.com/spacemeshos/go-spacemesh/priorityq"
+	"gopkg.in/tomb.v2"
+)
+
+// AtxProtocol is the protocol id for broadcasting atxs over gossip
+const AtxProtocol = "AtxGossip"
+
+type atxProcessor interface {
+	HandleGossipAtx(context.Context, service.GossipMessage, service.Fetcher)
+}
+
+// ATXListener handles ATX gossip messages.
+type ATXListener struct {
+	startOnce   sync.Once
+	tomb        tomb.Tomb
+	logger      log.Log
+	processor   atxProcessor
+	fetcher     service.Fetcher
+	atxMessages chan service.GossipMessage
+}
+
+// NewATXListener returns a instance of ATXListener
+func NewATXListener(net service.Service, processor atxProcessor, fetcher service.Fetcher, log log.Log) *ATXListener {
+	return &ATXListener{
+		processor:   processor,
+		fetcher:     fetcher,
+		logger:      log,
+		atxMessages: net.RegisterGossipProtocol(AtxProtocol, priorityq.Low),
+	}
+}
+
+// Start starts listening to ATX gossip messages.
+func (al *ATXListener) Start() {
+	al.startOnce.Do(func() {
+		al.tomb.Go(al.loop)
+	})
+}
+
+// Stop shutdown the listener gracefully.
+func (al *ATXListener) Stop() error {
+	al.logger.Info("stopping ATX gossip processing")
+	al.tomb.Kill(nil)
+	err := al.tomb.Wait()
+	al.logger.With().Info("stopped AX gossip processing with err", log.Err(err))
+	return err
+}
+
+// loop listens to gossip messages until ctx is canceled.
+func (al *ATXListener) loop() error {
+	al.logger.Info("start listening to ATX gossip")
+	for {
+		select {
+		case atx := <-al.atxMessages:
+			al.tomb.Go(func() error {
+				al.processor.HandleGossipAtx(al.tomb.Context(nil), atx, al.fetcher)
```

Comment:
i think this needs to be limited (e.g. after max number of handlers are spawned loop should either drop the ATX or block until the worker is available). if I read it right the previous code was always blocking until the previous ATX was Handled, but unlimited number of workers is also not a correct approach

Reply:
This is probably not the right place for this discussion, but I don't mind explaining it again later.

I think that we do a bad job of managing concurrency in the node. We don't currently have a way to prioritize urgent tasks (e.g. participating in Hare) over less urgent ones (e.g. processing an incoming ATX). The closest thing we have is a priority queue for gossip, which I think is not enough and we see the effects when running nodes on weak hardware on the testnet and tests.

I think there are clear priorities and those change through the course of the layer. Here's a rough outline:
![image](https://user-images.githubusercontent.com/776358/126963311-62902979-9d52-47d7-befe-a8c9d3df2508.png)

So the challenge is transitioning from having each module manage its own workers that consume an internal queue of tasks to code that manages centralized (within the node) queue of tasks and prioritizes them dynamically and a set of workers that consume tasks from the queue. The number of workers would be configurable, but by default it can be chosen by the node based on the machine its running on.

I believe this would considerably reduce the variability of the node's resource consumption and as a major added bonus, it would make debugging issues much easier since we can expose which tasks are being handled at any given moment making rationalizing what's going on in the node much easier (I imagine a "task log" with "task X started" and "task X completed" type messages).

Reply:
@noamnelke would you like to capture this in an issue, or have you already?

Reply:
it is hard to tell for me if it is a good idea or not, i believe it requires careful design. i also would like to find any good project in golang that is doing goroutine prioritization and runs well.

but as for this change, cpu is mostly idle on miners with an exception is actually handling atxs (you can see that in pyroscope history). so this is clearly a cpu bound task and we shouldn't blindly allocate a worker for every received atx. i suggest to limit number of workers that can process atx to NumCPU/2. 
![Screenshot from 2021-08-07 09-09-15](https://user-images.githubusercontent.com/1964576/128590169-f81280f3-abfd-411e-aec7-de117fc2cf95.png)

Reply:
I haven't yet. I've given it some thought and created this illustration, but I still need to sit down and dive into the details. If I won't get to it soon, I'll just start with what I've got and flesh it out later.